util/ranger/points.go

```diff
@@ -421,15 +422,48 @@ func (r *builder) buildFromNot(expr *expression.ScalarFunction) []point {
 	case ast.IsFalsity:
 		return r.buildFromIsFalse(expr, 1)
 	case ast.In:
-		rangePoints, hasNull := r.buildFromIn(expr)
+		var unsignedInt bool
+		if x, ok := expr.GetArgs()[0].(*expression.Column); ok {
+			unsignedInt = mysql.HasUnsignedFlag(x.RetType.Flag) && mysql.IsIntegerType(x.RetType.Tp)
+		}
+		inPoints, hasNull := r.buildFromIn(expr)
 		if hasNull {
 			return nil
 		}
-		retRangePoints := make([]point, 0, len(rangePoints)+2)
+		var rangePoints []point
+		// negative ranges can be directly ignored for unsigned int columns.
+		// Point `0` should be considered specially here. If the `IN` predicate contains `0`, `0` should be
+		// excluded from the result range, otherwise `0` should be included.
+		if unsignedInt {
+			rangePoints = make([]point, 0, 2+len(inPoints))
+			for i := 0; i < len(inPoints); i += 2 {
+				value := inPoints[i].value.GetInt64()
+				if value < 0 {
+					continue
+				} else if value == 0 {
+					rangePoints = append(rangePoints, point{value: types.NewIntDatum(0), start: true})
+					rangePoints = append(rangePoints, point{value: types.NewIntDatum(0)})
+				} else {
+					if len(rangePoints) == 0 {
+						rangePoints = append(rangePoints, point{value: types.NewIntDatum(0), start: true, excl: true})
```

Comment:
```
mysql> explain select * from t where a not in(111);
+---------------+---------+----------+------+--------------------------------------------------------------------------+---------+
| id            | parents | children | task | operator info                                                            | count   |
+---------------+---------+----------+------+--------------------------------------------------------------------------+---------+
| IndexScan_8   |         |          | cop  | table:t, index:a, range:(<nil>,0], (0,111), (111,+inf], keep order:false | 3833.33 |
| IndexReader_9 |         |          | root | index:IndexScan_8                                                        | 3833.33 |
+---------------+---------+----------+------+--------------------------------------------------------------------------+---------+
2 rows in set (0.00 sec)
```
When the column is `unsigned`. `(<nil>,111)` is the same as `(<nil>,0], (0,111)`.

Reply:
yes. I'll refine here